lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImpl.java

```diff
@@ -0,0 +1,84 @@
+package org.sagebionetworks.repo.model.dbo.schema;
+
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_JSON_SCHEMA_CREATED_BY;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_JSON_SCHEMA_CREATED_ON;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_JSON_SCHEMA_ID;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_JSON_SCHEMA_NAME;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_JSON_SCHEMA_ORG_ID;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.TABLE_JSON_SCHEMA;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+import org.sagebionetworks.ids.IdGenerator;
+import org.sagebionetworks.ids.IdType;
+import org.sagebionetworks.repo.model.schema.SchemaInfo;
+import org.sagebionetworks.repo.transactions.WriteTransaction;
+import org.sagebionetworks.util.ValidateArgument;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class JsonSchemaDaoImpl implements JsonSchemaDao {
+
+	@Autowired
+	private IdGenerator idGenerator;
+	@Autowired
+	private JdbcTemplate jdbcTemplate;
+
+	public static final RowMapper<SchemaInfo> SCHEMA_INFO_MAPPER = new RowMapper<SchemaInfo>() {
+
+		@Override
+		public SchemaInfo mapRow(ResultSet rs, int rowNum) throws SQLException {
+			SchemaInfo info = new SchemaInfo();
+			info.setNumericId(rs.getString(COL_JSON_SCHEMA_ID));
+			info.setOrganizationId(rs.getString(COL_JSON_SCHEMA_ORG_ID));
+			info.setName(rs.getString(COL_JSON_SCHEMA_NAME));
+			info.setCreatedBy(rs.getString(COL_JSON_SCHEMA_CREATED_BY));
+			info.setCreatedOn(rs.getTimestamp(COL_JSON_SCHEMA_CREATED_ON));
+			return info;
+		}
+	};
+
+	@WriteTransaction
+	@Override
+	public SchemaInfo createSchemaIfDoesNotExist(SchemaInfo schemaRoot) {
+		ValidateArgument.required(schemaRoot, "SchemaInfo");
+		ValidateArgument.required(schemaRoot.getOrganizationId(), "schema.organizationId");
+		ValidateArgument.required(schemaRoot.getName(), "schema.name");
+		ValidateArgument.required(schemaRoot.getCreatedBy(), "schema.createdBy");
+		ValidateArgument.required(schemaRoot.getCreatedOn(), "schema.createdOn");
+
+		// By inserting a row without an ID we ensure concurrent calls lock on the new
+		// row.
+		int updateCount = jdbcTemplate.update(
```

Comment:
We might be able to avoid having a nullable id while using the select for update first and still use the insert ignore below with a new id from the id generator.